### PR TITLE
Performance: contact searching by subqueries

### PR DESCRIPTION
**Pull request by rowanseymour**
**Simple vs complex queries**

Current search implementation tries to parse queries as complex, e.g. `a=x AND b=y`, and if that fails (e.g. user searches for just `joe`), runs it as a simple query on `name, urns__path, id` - implemented same way smartmin does its list view searches.

New search implementation uses a grammar which can parse both simple and complex syntaxes. A text token `x` is interpreted as the condition `*=x` where * is a placeholder for name/URN/id. Two adjacent conditions `x y` are implicitly combined with AND. 

For example `bob age>18` is parsed as `AND(*=bob, age>18)`

**URN querying**

Current search implementation matches a condition like `tel=x` or `twitter=y` against `urns__path` without checking the scheme. New search implementation will only match `twitter=y` if `scheme=twitter AND path=y`

**Optimisation**

A query like `a=x OR b=y OR a=z` will be parsed as `OR(OR(a=x, b=y), a=z)` as logical operators with same precedence are evaluated from left to right. This could be evaluated as 3 subqueries on the values table, but we introduce two steps to optimise this:

1. Flatten based on associativity of AND/OR: `OR(OR(a=x, b=y), a=z)` &rarr; `OR(a=x, b=y, a=z)`
2. Regroup conditions on the same property: `OR(a=x, b=y, a=z)` &rarr; `OR(OR(a=x, a=z), b=y)`

During queryset generation we can then treat `OR(a=x, a=z)` as a special case and implement it as a single subquery on `Value` rather than two.

We can also further optimise `OR(a=x, a=z)` by implementing it as `a IN (x, y)` which Postgres will evaluate with a single index scan on the values table. This should be particularly beneficial for a lot of dynamic group queries which are like `district=A or district=B ... or district=N`

**Dynamic group membership checks**

The subquery approach is less efficient in the case where we want to check whether a single contact matches the search criteria. We address this by passing the contact being checked down to each condition so that it can be explicitly filtered on in the subquery. I've checked that this retains the performance gains from #1150 and also proposed #1156 to report in sentry if any membership check takes too long.

**Error handling**

Current implementation reverts to a simple search (which can't error) if a complex search generates an error. This implementation lets search errors propagate up to the view level so that they could be exposed in the UI. It also hides the "Save As Group" button so that an invalid search query can't be saved as a dynamic group.

screen shot 2017-03-09 at 10 23 47

screen shot 2017-03-09 at 10 24 54

### CODE REVIEW

File: temba/contacts/search.py [hunk 1]
```diff
@@ -56,246 +26,442 @@ def __init__(self, message):
         self.message = message
 
 
-def contact_search(org, query, base_queryset):
+class SearchLexer(object):
```
Comment (rowanseymour) [inline temba/contacts/search.py]: Making the lexer a class makes it easier to pass around state without resorting to globals

File: pip-freeze.txt [hunk 2]
```diff
@@ -76,7 +76,7 @@ phonenumbers
 Pillow==3.4.2
 pisa==3.0.33
 plivo==0.10.1
-ply==3.4
+ply==3.10
```
Comment (rowanseymour) [inline pip-freeze.txt]: No major changes - just bug fixes and Python 3 stuff

File: temba/api/tests/test_v1.py [hunk 3]
```diff
@@ -1937,6 +1937,7 @@ def test_api_contact_actions(self):
         test_contact = Contact.get_test_contact(self.user)
 
         group = self.create_group("Testers")
+        self.create_field('isdeveloper', "Is developer")
```
Comment (rowanseymour) [inline temba/api/tests/test_v1.py]: An exception is raised now if you try to create a dynamic group based on non-existent fields. We shield users from this by not letting them create a group in the first place if their search contains errors.

File: temba/api/tests/test_v2.py [hunk 4]
```diff
@@ -1060,7 +1060,7 @@ def test_contacts(self):
         self.assertEqual(set(jaqen.user_groups.all()), set())
         self.assertEqual(set(jaqen.values.all()), set())
 
-        dyn_group = self.create_group("Dynamic Group", query="nickname has Ja")
+        dyn_group = self.create_group("Dynamic Group", query="nickname is jado")
```
Comment (rowanseymour) [inline temba/api/tests/test_v2.py]: Text fields can no longer be searched with has/~

File: temba/contacts/models.py [hunk 5]
```diff
@@ -964,17 +964,16 @@ def make_urn(tel_as_int):
...
-        or not the query was a valid complex query, e.g. name = "Bob" AND age = 21
+        Performs a search of contacts within a group (system or user)
         """
-        from temba.contacts import search
+        from .search import contact_search
 
-        if base_queryset is None:
-            base_queryset = Contact.objects.filter(org=org, is_active=True, is_test=False, is_blocked=False, is_stopped=False)
+        if not base_group:
+            base_group = ContactGroup.all_groups.get(org=org, group_type=ContactGroup.TYPE_ALL)
```
Comment (rowanseymour) [inline temba/contacts/models.py]: All contact searching is now based on a group even if it's a system group like "All Contacts". In current implementation, searching in the UI is always based on a group, but dynamic group membership checks don't use a group.

File: temba/values/migrations/0010_value_indexes.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+CREATE INDEX values_value_field_datetime_value_not_null
+ON values_value(contact_field_id, datetime_value)
+WHERE contact_field_id IS NOT NULL AND datetime_value IS NOT NULL;
+
+CREATE INDEX values_value_field_decimal_value_not_null
+ON values_value(contact_field_id, decimal_value)
+WHERE contact_field_id IS NOT NULL AND decimal_value IS NOT NULL;
+
+ANALYZE values_value;
```
Comment (rowanseymour) [inline temba/values/migrations/0010_value_indexes.py]: Not sure if this belongs here but basically we need ANALYZE to be run for this indexes to assist with statistics. Without putting this here, the performance tests which run on a clean database will be slower than expected.

File: temba/contacts/search.py [hunk 7]
```diff
@@ -1,311 +1,538 @@
...
-    except SearchException:
-        pass
+    literals = '()'
 
-    # if that didn't work, try again as simple name or urn path query
-    return contact_search_simple(org, query, base_queryset), False
+    reserved = {
+        'and': 'AND',
+        'or': 'OR',
+        'has': 'COMPARATOR',
```
Comment (ericnewcomer) [inline temba/contacts/search.py]: Didn't we talk about removing has?

File: temba/values/migrations/0010_value_indexes.py (hunk 6 above)
Comment (ericnewcomer) [inline temba/values/migrations/0010_value_indexes.py]: Shouldn't these be CONCURRENTLY?

File: temba/contacts/search.py (hunk 7 above)
Comment (rowanseymour) [inline temba/contacts/search.py]: We need to keep it around for name and URN queries - but we've removed it for contact fields, except for location fields. I'm not convinced it's useful to keep for location fields but someone else felt pretty strongly it was :)

File: temba/values/migrations/0010_value_indexes.py (hunk 6 above)
Comment (rowanseymour) [inline temba/values/migrations/0010_value_indexes.py]: Yup

File: temba/contacts/search.py (hunk 7 above)
Comment (nicpottier) [inline temba/contacts/search.py]: I have no idea who that could be but they sound wise and handsome.

**Pull request merged**